### ISSUE TITLE
envconsul fails to populate environment variables for ruby on rails console

### ISSUE THREAD

**Issue by number61971** [hashicorp/envconsul#308]
There appears to be a breaking change between versions 0.11.0 and 0.12.0+. It happens when the `-once` command line option is specified. Our attempt to instantiate a Ruby on Rails console—which depends upon values being available in environment variables populated by `envconsul`—simply freezes in place indefinitely. There is no logging/debugging output to help us.

Here's the config file we use (`/usr/local/etc/envconsul.hcl`).

```
vault {
  renew_token = false
  retry {
    attempts = 3
  }
}
secret {
  format = "DB_PASSWORD"
  path = "secret/db_password/prod"
  no_prefix = true
}
secret {
  format = "MESSAGE_QUEUE_PASSWORD"
  path = "secret/message_queue_password/prod"
  no_prefix = true
}
secret {
  format = "AWS_SECRET_ACCESS_KEY"
  path = "secret/aws_secret_access_key/prod"
  no_prefix = true
}
secret {
  format = "AWS_ACCESS_KEY_ID"
  path = "secret/aws_access_key_id/prod"
  no_prefix = true
}
```

And here are the commands we're trying to run

```
$ export VAULT_ADDR="https://$VAULT_HOST:8200"
$ export VAULT_SKIP_VERIFY=1
$ export VAULT_TOKEN=$(vault login -method=aws -token-only=true header_value=$VAULT_HOST role=$VAULT_ROLE)
$ envconsul -once -upcase -config=/usr/local/etc/envconsul.hcl "rails c"
```

This process has worked for all versions of `envconsul` between 0.7.3 and 0.11.0, but is silently broken for all versions >= 0.12.0.

Here's what this looks like when successful:

```
$ envconsul -once -upcase -config=/usr/local/etc/envconsul.hcl "rails c"
2022/08/02 18:58:20.570297 [WARN] (clients) disabling vault SSL verification
=> Booting Rails
Loading production environment (Rails 6.1.6.1)
irb(main):001:0>
```

And here's what it looks like when it is silently failing, freezing in place:

```
envconsul -once -upcase -config=/usr/local/etc/envconsul.hcl "rails c"
2022-08-02T19:00:33.417Z [WARN] (clients) disabling vault SSL verification
=> Booting Rails
```

**Comment by eikenb** [comment]
Hey @number61971, thanks for reporting this!

TLDR; to get this working fast...

Either wrap 'rails c' in a (very) short script and call it from the envconsul command line..

```sh
#!/bin/sh
rails c
```
Or put it in the config file as an exec segment like so...

```
exec {
    command = ["rails", "c"]
}
```

--

Long winded explanation.

What you are seeing here is an issue that boils down to command parsing. Envconsul early on adopted the idea of trying to parse commands like a shell would and this didn't work out in the long run. To try to address the issues with the command parsing we had to stop trying to parse like a shell and, to maintain compatibility, we used a shell to parse things. That is we took commands that would require parsing and passed them as a string to an `sh -c 'command string here'` but this had the issue of the wrapping shell intercepting some OS Signals and to maintain compatibility with that I needed to use the system `setgpid` (set process group id). This has the interesting side effect of breaking interactivity.

In envconsul 0.13.0 (via consul-template 0.29.2) fixed this as much as possible by using setgpid ONLY when `sh -c` is used. Which is only used used when the command is a free form string with spaces present. If you use a single command (shell script) it won't set it and if you use the exec block you can define the command in a pre-parsed list format (eg. above).

If you have any questions or ideas on how to improve this please let me know. I should probably make sure the documentation has been updated to reflect this better in addition to anything else.

Hope this helps.

**Comment by eikenb** [comment]
> If you use a single command (shell script) it won't set it and if you use the exec block you can define the command in a pre-parsed list format (eg. above).

Re-reading this today it seems not very clear. What I meant was that it doesn't call `setpgid` in either of those cases.

**Comment by number61971** [comment]
Hi @eikenb,

Thanks so much for the reply! Sorry it's taken me forever to get back to this issue and test your solution.

I ended up creating a little shell script wrapping the "rails c" invocation, and it worked perfectly, exactly as you suggested.

Thank you very much for the help.
